### PR TITLE
[MODSOURCE-530] Send DuplicateException if incoming file contains duplicate records

### PR DESCRIPTION
**Pull request by RomanChernetskyi**
## Purpose
Duplicate records in incoming file causes problems after overlay process with no error reported

## Approach
Changed a way of assigning matched id, now it generated during save record process, matched id is retrieved from 999$s field, in case if the parsed record does not contain this field, assigning id of exiting source record with a same external id

### CODE REVIEW

File: mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java [hunk 1]
```diff
@@ -250,6 +275,50 @@ public Future<Void> updateRecordsState(String matchedId, RecordState state, Reco
     return recordDao.updateRecordsState(matchedId, state, recordType, tenantId);
   }
 
+  private Future<Record> setMatchedIdForRecord(Record record, String tenantId) {
```
Comment (okolawole-ebsco) [inline mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java:278]: please add some comments to the body of this method.

File: mod-source-record-storage-server/src/main/java/org/folio/services/util/AdditionalFieldsUtil.java [hunk 2]
```diff
@@ -587,10 +599,12 @@ && isNotBlank(record.getParsedRecord().getContent().toString())) {
...
-        MarcReader reader = buildMarcReader(record);
-        if (reader.hasNext()) {
-          return reader.next();
-        }
+        try {
+          MarcReader reader = buildMarcReader(record);
+          if (reader.hasNext()) {
+            return reader.next();
+          }
+        } catch (Exception ignored) {}
```
Comment (okolawole-ebsco) [inline mod-source-record-storage-server/src/main/java/org/folio/services/util/AdditionalFieldsUtil.java:607]: why doesn't it make sense to log the exception?

File: mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java (hunk 1 above)
Comment (RomanChernetskyi) [reply mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java:278]: done, thanks

File: mod-source-record-storage-server/src/main/java/org/folio/services/util/AdditionalFieldsUtil.java (hunk 2 above)
Comment (RomanChernetskyi) [reply mod-source-record-storage-server/src/main/java/org/folio/services/util/AdditionalFieldsUtil.java]: done, thanks

File: mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java [hunk 3]
```diff
@@ -250,6 +275,54 @@ public Future<Void> updateRecordsState(String matchedId, RecordState state, Reco
...
+          } else {
+            LOG.warn("setMatchedIdForRecord:: Error while retrieving source record");
+            promise.fail(ar.cause());
+          }
+        });
+    } else {
+      // Set matched id same as record id
+      promise.complete(record.withMatchedId(record.getId()));
+    }
+    return promise.future().onSuccess(r -> addFieldToMarcRecord(r, TAG_999, SUBFIELD_S, r.getMatchedId()));
```
Comment (VRohach) [inline mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java:314]: @RomanChernetskyi  Is there an opportunity to divide this method via other methods? The body in this "if" condition seems so complex. Thank you in advance.

**Review by VRohach — COMMENTED** [review]
@RomanChernetskyi  Update NEWS please.

File: mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java (hunk 3 above)
Comment (RomanChernetskyi) [reply mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java]: Done, thanks

**Review by VRohach — APPROVED** [review]

**Review by yaroslav-epam — APPROVED** [review]

**Pull request merged**